### PR TITLE
Add advanced password generator features

### PR DESCRIPTION
**Pull request by seregaxvm**
## Description

Added options to increase control over the character groups used for password generation.

Special characters group has been split into 6 subgroups (see screenshot below). Buttons controlling these groups and extended ASCII button were moved to fineTuneBar widget and are visible only when Fine Tune button is pressed. 

## Motivation and context
A lot of sites have very particular set of characters allowed in password. This change allows to generate a password in compliance with character restrictions simply by selecting character groups.

fineTuneBar is initially hidden so additional information will not overload the UI.

Extended ASCII button was moved inside fineTuneBar because firstly, it makes sense, secondly, so UI would not change a lot. 

## How has this been tested?

- It has been tested by hand.
- Automated tests were run and passed.
- Program was build and run with `-DWITH_ASAN=ON` before and after alterations. No new memory leaks were introduced (BTW there are some memory leaks on development brunch).

### Before changes
selection_206

### After changes
selection_207

### After changes with fine tune button pressed
selection_208

## Types of changes

- ✅ New feature (non-breaking change which adds functionality)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

**Comment by droidmonkey** [comment]
Oh i like this, although i think a label of "Advanced..." or "Extended ASCII..." would be more appropriate.

**Comment by seregaxvm** [comment]
Replaced FineTune button with Advanced button on the side. When clicked, it reveals advanced configurations. User can get back to simple configurations by clicking Simple button.

Added ability to create hexadecimal passwords, proposed in #1248 by splitting characters is subgroups.

### Simple mode
selection_209
### Advanced mode
selection_210

**Comment by TheZ3ro** [comment]
I really like this new UI, I will review in the next few days

**Comment by glubsy** [comment]
What if I only want some specific characters among `<*+!?=` and only some characters among `\_|-/` (as required by some programs / websites)?

**Comment by seregaxvm** [comment]
If it is a frequent issue, groups should be adjusted. If it is just a couple of sites/programs, manual password edit is fine. I don't think adding more buttons would be a good idea. It would be too messy.

**Comment by glubsy** [comment]
But what if I generate a lot of passwords, and they need to be 500+ characters? 

Wouldn't a filter be more appropriate? Like a blacklist / whitelist pair or small input fields perhaps?

**Comment by droidmonkey** [comment]
There was discussion of including a textbox where a white/black list could be applied.

**Comment by seregaxvm** [comment]
Do you need this in GUI? Could this just be a CLI feature?

**Comment by droidmonkey** [comment]
Yes it absolutely must be in the GUI

**Comment by seregaxvm** [comment]
selection_234

selection_235

@droidmonkey @glubsy I've added "do not include" text input to gui and cli.
Buttons have precedence over this input.  If all characters form some group are mentioned in "do not include" field, then all characters from this group may appear in generated password. It is done like that to ensure that there's no inconsistency between selected groups and generated password.

**Comment by glubsy** [comment]
@seregaxvm this is looking great in my opinion! Thank you!

**Comment by droidmonkey** [comment]
Those are for hex passwords.

I don't understand the purpose of overriding the "do not include" entries if it covers the whole spectrum of a button. My opinion is if you put it in the do not include box, you mean it. 

I would add a clear button similar to what we have in the search box to quickly remove the do not include entries.

If you hide the advanced fields, does it revert back to normal behavior?

**Comment by seregaxvm** [comment]
selection_236

selection_237

@droidmonkey  @glubsy
- Hexadecimal character buttons removed. One can generate hexadecimal passwords using "do not include" field. This is not very convenient but, as I understand, this feature is not so frequently used anyway.
- Clear button added to "do not include" field.
- "do not include" entries only applied when the field is visible. They are ignored in simple mode.
- "do not include" has precedence over buttons. Characters from this field are never in the password.
- When all characters, selected with buttons, are in "do not include" field, input is invalidated and no password will be generated. 
- Added `--exclude_similar` and `--every_group` options to CLI generator.
- Conversion `toLatin1` is used to calculate entropy so Unicode characters were not added.

**Comment by glubsy** [comment]
Well this is looking very good. 

Perhaps the only thing that might confuse the user is when clicking `Advanced`, the `a-z` button is shifting below, and the `0-9` button is shifting left.

Also it might be a good opportunity to add examples in the `Exclude look-alike characters`. I, for one, don't know which characters are targeted by this.

**Comment by droidmonkey** [comment]
Awesome update!

**Comment by seregaxvm** [comment]
selection_039

@glubsy added tool-tip description with list of excluded look-alike characters.
The last one is from Extended ASCII group. Should "." from special group also be excluded?
`a-z`, `A-Z` and `0-9` are grouped like so, firstly, to move related buttons closer to each other, secondly, to have columns of buttons with roughly same width.

**Comment by glubsy** [comment]
Fantastic! Thank you very much!

**Comment by droidmonkey** [comment]
I tried this out and it worked perfectly. Recommend the following additions/changes:

adv_pwgen

The HEX button (oft requested feature) would very simply set the "exclude characters" textbox to ```GHIJKLMNOPQRSTUVWXYZghijklmnopqrstuvqrxyz``` it is up to the user to select the A-Z (or a-z) and 0-9 buttons.

**Comment by seregaxvm** [comment]
@droidmonkey, please create branch `feature/finetune-password` so I could rebase to it?

**Comment by seregaxvm** [comment]
keepassxc_238
keepassxc_239

Checkboxes are hidden in simple mode. Added `Hex` shortcut. When clicked, it fills "exclude characters" textbox with predefined string `GHIJKLMNOPQRSTUVWXYZghijklmnopqrstuvwxyz`

**Comment by droidmonkey** [comment]
@seregaxvm you create the branch on your own fork.

**Comment by seregaxvm** [comment]
@droidmonkey  I did but I'm not able to change the brunch to pull from. I only can change base brunch to commit into, but it should exist.

**Comment by droidmonkey** [comment]
Crap true, you cannot change the base branch. That is fine, I can merge it right now without much issue. In the future make sure you make  PR using a branch off develop.

File: tests/TestCryptoHash.cpp [hunk 1]
```diff
@@ -50,8 +50,9 @@ void TestCryptoHash::test()
                                  "8d2877eec2f63b931bd47417a81a538327af927da3e"));
 
     QByteArray result3 = CryptoHash::hash(source2, CryptoHash::Sha512);
-    QCOMPARE(result3, QByteArray::fromHex("0d41b612584ed39ff72944c29494573e40f4bb95283455fae2e0be1e3565aa9f48057d59e6ff"
-                                          "d777970e282871c25a549a2763e5b724794f312c97021c42f91d"));
+    QCOMPARE(result3,
```
Comment (droidmonkey) [inline tests/TestCryptoHash.cpp]: Revert these changes, your code formatter got a little happy.

File: tests/TestCsvExporter.cpp [hunk 2]
```diff
@@ -64,9 +64,10 @@ void TestCsvExporter::testExport()
     QVERIFY(buffer.open(QIODevice::ReadWrite));
     m_csvExporter->exportDatabase(&buffer, m_db);
 
-    QString expectedResult =
-        QString().append(ExpectedHeaderLine).append("\"Test Group Name\",\"Test Entry Title\",\"Test Username\",\"Test "
-                                                    "Password\",\"http://test.url\",\"Test Notes\"\n");
+    QString expectedResult = QString()
```
Comment (droidmonkey) [inline tests/TestCsvExporter.cpp]: Revert these changes, your code formatter got a little happy.

File: tests/TestKdbx4.cpp [hunk 3]
```diff
@@ -317,11 +317,11 @@ void TestKdbx4::testUpgradeMasterKeyIntegrity_data()
     QTest::newRow("Upgrade (explicit): kdf-argon2") << QString("kdf-argon2") << KeePass2::FILE_VERSION_4;
     QTest::newRow("Upgrade (explicit): kdf-aes-kdbx4") << QString("kdf-aes-kdbx4") << KeePass2::FILE_VERSION_4;
     QTest::newRow("Upgrade (implicit): public-customdata") << QString("public-customdata") << KeePass2::FILE_VERSION_4;
-    QTest::newRow("Upgrade (implicit): rootgroup-customdata") << QString("rootgroup-customdata")
-                                                              << KeePass2::FILE_VERSION_4;
+    QTest::newRow("Upgrade (implicit): rootgroup-customdata")
```
Comment (droidmonkey) [inline tests/TestKdbx4.cpp]: Revert these changes, your code formatter got a little happy.

File: tests/TestKdbx4.cpp [hunk 4]
```diff
@@ -317,11 +317,11 @@ void TestKdbx4::testUpgradeMasterKeyIntegrity_data()
...
+        << QString("rootgroup-customdata") << KeePass2::FILE_VERSION_4;
     QTest::newRow("Upgrade (implicit): group-customdata") << QString("group-customdata") << KeePass2::FILE_VERSION_4;
-    QTest::newRow("Upgrade (implicit): rootentry-customdata") << QString("rootentry-customdata")
-                                                              << KeePass2::FILE_VERSION_4;
+    QTest::newRow("Upgrade (implicit): rootentry-customdata")
```
Comment (droidmonkey) [inline tests/TestKdbx4.cpp]: Revert these changes, your code formatter got a little happy.

**Comment by seregaxvm** [comment]
@droidmonkey 
> Revert these changes, your code formatter got a little happy.

Done
> In the future make sure you make PR using a branch off develop.

Ok, sorry about that.

**Comment by phoerious** [comment]
I'm confused. You can change the base branch at the top of this page.

**Comment by seregaxvm** [comment]
Yes, I can change it, but only to branch that already exists. I have no write access, so I cannot create `feature/finetune-password` branch. If you would create it, I will be able to change base branch of this pr to `feature/finetune-password`.

**Comment by droidmonkey** [comment]
I can change the target branch (ie develop) but not the base, GitHub doesnt even give me the option to try.

**Comment by seregaxvm** [comment]
selection_065

These are all my options. If it's possible I would prefer not to complicate things and to merge into `develop`. If it's not desirable, I will open another pull request.

**Comment by TheZ3ro** [comment]
I really like this PR and the new UI (my 2 cent)

**Comment by droidmonkey** [comment]
This is ready for merge. Any more changes requested? I'm very happy with the results as well.

File: src/core/PasswordGenerator.cpp [hunk 5]
```diff
@@ -220,6 +285,29 @@ QVector<PasswordGroup> PasswordGenerator::passwordGroups() const
         passwordGroups.append(group);
     }
 
+    int i = 0;
+    while (1) {
```
Comment (TheZ3ro) [inline src/core/PasswordGenerator.cpp]: It's not very clear at first look what this piece of code is suppose to do, insert a comment or move it into a function

Comment (seregaxvm) [reply src/core/PasswordGenerator.cpp]: How about that :"Loop over character groups and remove excluded characters from them; remove empty groups"? Also break condition should be moved from `if` to `while`

Comment (droidmonkey) [reply src/core/PasswordGenerator.cpp]: Also, you should use while(true) instead. You are implicitly converting 1 to bool in this case.

Comment (TheZ3ro) [reply src/core/PasswordGenerator.cpp]: @seregaxvm I think that can do it. :wink:

Comment (seregaxvm) [reply src/core/PasswordGenerator.cpp]: done

**Comment by droidmonkey** [comment]
I will do a formal review very soon

**Comment by Gittyperson** [comment]
It's great that these features are coming. Many, many sites accept the password during registration (containing the more unusual, non-allowed characters) but when it's time to log in, they will report "password is wrong" with no further details.

**Comment by lordloh** [comment]
I would like to suggest adding a check box called "SQL safe" that would remove things like "--" and ";" from the password. I am not sure how developers handle them. I have had trouble logging into websites after using a password generated by Keepassxc. The password is accepted at change password time, but not during login :-( 

I had to tweak out certain sets of characters to get it working. This is completely the websites fault. But it would be ages before they fix it. I discovered at one site that the desktop site accepts longer passwords than the mobile version of the site. I had to use a shorter password :-(

**Comment by droidmonkey** [comment]
Omg, if you are running into websites that can't handle those characters that means they are ripe for SQL injection attacks. RUN!

**Comment by lordloh** [comment]
I am inclined to think that different sections section of their sites are
handling SQL injections differently.

Perhaps the order of sanitization and hashing passwords are different(?)

**Comment by TheZ3ro** [comment]
> You can just type `;-)` or something like that into the `do not include` line.

The wink emoji of SQLi doom

**Comment by sebast889** [comment]
@seregaxvm @droidmonkey Sorry I couldn't find anywhere the reason so I'm asking here. Why are the character types in the group that they are in? Seems they are arbitrarily chosen or at least I couldn't spot any real pattern. Is there any rationale behind what characters are in what group?

**Comment by droidmonkey** [comment]
This is already merged. If you have a suggested grouping(s) please out them in a new issue.

**Comment by seregaxvm** [comment]
@sebast889 They are based on the personal experience with different sites password restrictions. The idea was to split `Special Characters` group into smaller ones. Their names are `Braces`, `Punctuation`, `Quotes`, `Math`, `Dashes` and `Logograms`.
